packages/@expo/cli/src/install/installExpoPackage.ts

```diff
@@ -0,0 +1,74 @@
+import * as PackageManager from '@expo/package-manager';
+import { spawn } from 'child_process';
+
+import * as Log from '../log';
+import { getRunningExpoProcessesForDirectory } from '../utils/getRunningProcess';
+import { isInteractive } from '../utils/interactive';
+import { confirmAsync } from '../utils/prompts';
+
+/**
+ * Given a list of incompatible packages, installs the correct versions of the packages with the package manager used for the project.
+ * This exits immediately after spawning the install command, since the command shouldn't remain running while it is being updated.
+ */
+export async function installExpoPackageAsync(
+  projectRoot: string,
+  {
+    packageManager,
+    packageManagerArguments,
+    expoPackageToInstall,
+    followUpCommand,
+  }: {
+    /** Package manager to use when installing the versioned packages. */
+    packageManager: PackageManager.NodePackageManager;
+    /**
+     * Extra parameters to pass to the `packageManager` when installing versioned packages.
+     * @example ['--no-save']
+     */
+    packageManagerArguments: string[];
+    expoPackageToInstall: string;
+    followUpCommand: string | undefined;
+  }
+) {
+  const runningExpoProcesses = await getRunningExpoProcessesForDirectory(projectRoot);
+
+  if (runningExpoProcesses.length) {
+    Log.warn(
+      'The Expo CLI is running this project in another terminal window. It must be closed before installing.'
+    );
+    let killExpoProcesses = false;
+    if (isInteractive()) {
+      killExpoProcesses = await confirmAsync({ message: 'Kill Expo CLI processes?' });
+    }
+
+    if (killExpoProcesses) {
+      runningExpoProcesses.forEach((pid) => {
+        process.kill(pid);
+      });
+    } else {
+      Log.exit(
+        'Aborting install since the Expo CLI is running. Kill any open Expo CLI processes for this project and try again.',
+        1
+      );
+    }
+  }
+
+  // prevents the package manager from actually installing the package
+  // we will grab the command it intended to run and run it with the follow up command
+  packageManager.simulate = true;
+
+  await packageManager.addAsync([...packageManagerArguments, expoPackageToInstall]);
+
+  const expoInstallCommand = packageManager.lastCommand;
```

Comment:
could we make a method on package manager that returns the command instead of adding this state? i'm hesitant here about adding multiple instance variables to accomplish this somewhat awkwardly. could do this instead:

```
packageManager.getCommandString('addAsync', [...packageManagerArguments, expoPackageToInstall]);
```

Reply:
Just updated to encapsulate all of this in a separate function with no changes to the constructor params, etc. I called it `addDeferredAsync()`, restricting it to just adding packages. While there's no reason it couldn't be more flexible to defer any command, there's also no reason currently that it needs to be. 

I added the check to prevent a second execution from overwriting the last command and confirmed there isn't an infinite install loop if the `expo` install command fails. There would likely still be problems if a future package manager implementation called `runAsync` multiple times for a single `addAsync` call. I suppose this could be solved by pushing commands to an array.

I'm still not thrilled about needing to feed `spawnAsync` a no-op (I also tried making a mock `SpawnResult` but struggled with either satisfying TS or getting the promises to resolve), but any variation of changing the return type of `runAsync` sends a pretty huge ripple through the individual `PackageManager` implementations.

If this whole approach is off, I think another solution could be to make each package manager implement its own `addDeferredAsync()` that basically reimplements `addAsync` without returning the `SpawnPromise`. I still prefer to make this work with only base class changes.